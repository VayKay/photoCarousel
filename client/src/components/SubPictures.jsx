/* eslint-disable react/prop-types */
import React from 'react';


class SubPictures extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      height: '50%',
      width: (window.innerWidth / 4),
      hoverTwo: false,
      hoverThree: false,
      position: '50%',
    };

    this.updateDimensions = this.updateDimensions.bind(this);
    this.hoverHandler = this.hoverHandler.bind(this);
  }

  componentDidMount() {
    window.addEventListener('resize', this.updateDimensions);
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.updateDimensions);
  }

  updateDimensions() {
    const { threshold } = this.props;
    if (window.innerWidth < threshold) {
      this.setState({
        width: (window.innerWidth / 4),
        height: (window.innerHeight / 4),
        position: '75%',
      });
    } else {
      this.setState({
        width: (window.innerWidth / 4),
        height: (window.innerHeight / 4),
        position: '50%',
      });
    }
  }

  hoverHandler({ target }) {
    const { hoverTwo, hoverThree } = this.state;
    const { hoverHandler } = this.props;
    const { name } = target;

    if (name === 'two') {
      if (!hoverTwo) {
        hoverHandler('two');
      } else {
        hoverHandler('clear');
      }
      this.setState({ hoverTwo: !hoverTwo });
    }

    if (name === 'three') {
      if (!hoverThree) {
        hoverHandler('three');
      } else {
        hoverHandler('clear');
      }
      this.setState({ hoverThree: !hoverThree });
    }
  }

  render() {
    const { height, width, position } = this.state;
    const {
      photos, totalHeight, totalWidth, opacityTwo, opacityThree,
    } = this.props;
    return (
      <div className="subPictureColumn" style={{ left: position, height: '60%' }}>
        <div className="img-hover-zoom">
          <div className="entryOne">
            <img
              src={photos[0]}
              id="subPictures"
              className="subPicture"
              alt="http://lorempixel.com/1440/960/city/"
              style={{
                top: 0, minHeight: (totalHeight * 0.3), opacity: opacityTwo,
              }}
              height={height}
              width={width}
              name="two"
              onMouseEnter={this.hoverHandler}
              onMouseLeave={this.hoverHandler}
            />
          </div>
        </div>
        <div className="img-hover-zoom">
          <div className="entryTwo">
            <img
              src={photos[1]}
              id="subPictures"
              className="subPicture"
              alt="http://lorempixel.com/1440/960/city/"
              style={{
                top: 0, minHeight: (totalHeight * 0.3), opacity: opacityThree,
              }}
              height={height}
              width={width}
              name="three"
              onMouseEnter={this.hoverHandler}
              onMouseLeave={this.hoverHandler}
            />
          </div>
        </div>
      </div>
    );
  }
}

export default SubPictures;
