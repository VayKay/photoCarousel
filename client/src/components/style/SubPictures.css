.imgHoverZoom {
    -webkit-box-sizing: border-box;
    -moz-box-sizing: border-box;
    box-sizing: border-box;
    -ms-flex: 0 1 auto;
}

.imgHoverZoom:hover img {
    transform: scale(1.05) !important;
    transition: -ms-transform 450ms cubic-bezier(0.645, 0.045, 0.355, 1) 0s, 
    -webkit-transform 450ms cubic-bezier(0.645, 0.045, 0.355, 1) 0s, transform 450ms 
    cubic-bezier(0.645, 0.045, 0.355, 1) 0s !important;
}

.blur {
    filter: blur(0.9em);
    opacity: 1 !important;
}


.subPictureColumn {
    flex-direction: column !important;
    position: relative !important;
    background-color: black !important;
    overflow: hidden !important;
    border: 1px solid;
    display: inline-block;
}

.entryOne {
    -webkit-box-pack: center !important;
    -webkit-box-align: center !important;
    display: flex !important;
    align-items: center !important;
    justify-content: center !important;
    vertical-align: middle !important;
    position: relative !important;
    overflow: hidden !important;
    transform: scale(1) !important;
    box-sizing: border-box;
}

.entryTwo {
    -webkit-box-pack: center !important;
    -webkit-box-align: center !important;
    display: flex !important;
    align-items: center !important;
    justify-content: center !important;
    vertical-align: middle !important;
    position: relative !important;
    overflow: hidden !important;
    border-top: 1px solid;
    box-sizing: border-box;
}


.subPicture {
    background-size: cover!important;
    background-position: center bottom!important;
    object-fit:cover;
    transition: filter 0.5s, -ms-transform 450ms cubic-bezier(0.645, 0.045, 0.355, 1) 0s, -webkit-transform 450ms cubic-bezier(0.645, 0.045, 0.355, 1) 0s, transform 450ms cubic-bezier(0.645, 0.045, 0.355, 1) 0s, opacity 450ms cubic-bezier(0.645, 0.045, 0.355, 1) 0s !important;
} 