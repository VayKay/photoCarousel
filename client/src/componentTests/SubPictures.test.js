import React from 'react';
import { shallow } from 'enzyme';
import SubPictures from '../components/SubPictures';

describe('SubPictures', () => {
  it('should render correctly', () => {
    const component = shallow(<SubPictures photos={['http://lorempixel.com/1440/960/city', 'http://lorempixel.com/1440/960/city']} tinyPhotos={['http://lorempixel.com/60/40/city', 'http://lorempixel.com/60/40/city']} />);

    component.setState({
      opacityThree: 1, opacityTwo: 1, threshold: 1008, totalHeight: 766, totalWidth: 1440,
    });

    expect(component).toMatchSnapshot();
  });

  it('should load sub pictures progressively on mount', () => {
    const spy = jest.spyOn(SubPictures.prototype, 'progressiveLoading');
    const component = shallow(<SubPictures photos={['http://lorempixel.com/1440/960/city', 'http://lorempixel.com/1440/960/city']} tinyPhotos={['http://lorempixel.com/60/40/city', 'http://lorempixel.com/60/40/city']} />);

    component.instance().progressiveLoading();
    expect(spy).toHaveBeenCalled();
  });

  it('should handle hover correctly', () => {
    const mockCallBack = jest.fn();
    const component = shallow(<SubPictures photos={['http://lorempixel.com/1440/960/city', 'http://lorempixel.com/1440/960/city']} hoverHandler={mockCallBack} tinyPhotos={['http://lorempixel.com/60/40/city', 'http://lorempixel.com/60/40/city']} />);

    component.setState({
      hoverTwo: false,
      hoverThree: false,
    });

    component.instance().hoverHandler({ target: { name: 'two' } });
    expect(component.state('hoverTwo')).toEqual(true);
    component.instance().hoverHandler({ target: { name: 'three' } });
    expect(component.state('hoverThree')).toEqual(true);
  });

  it('should handle hovering off correctly', () => {
    const mockCallBack = jest.fn();
    const component = shallow(<SubPictures photos={['http://lorempixel.com/1440/960/city', 'http://lorempixel.com/1440/960/city']} hoverHandler={mockCallBack} tinyPhotos={['http://lorempixel.com/60/40/city', 'http://lorempixel.com/60/40/city']} />);

    component.setState({
      hoverTwo: true,
      hoverThree: true,
    });

    component.instance().checkHover({ target: { name: 'two' } });
    expect(component.state('hoverTwo')).toEqual(false);
    component.instance().checkHover({ target: { name: 'three' } });
    expect(component.state('hoverThree')).toEqual(false);
  });

  it('should handle clicking on the second or third picture correctly', () => {
    const mockCallBack = jest.fn();
    const component = shallow(<SubPictures photos={['http://lorempixel.com/1440/960/city', 'http://lorempixel.com/1440/960/city']} hoverHandler={mockCallBack} onClickHandler={mockCallBack} tinyPhotos={['http://lorempixel.com/60/40/city', 'http://lorempixel.com/60/40/city']} />);
    component.setProps({ clickHandler: mockCallBack });

    component.instance().onClickHandler({ target: { name: 'two' } });
    component.instance().onClickHandler({ target: { name: 'three' } });
    expect(mockCallBack.mock.calls.length).toEqual(2);
  });


  it('should handle window resize', () => {
    const component = shallow(<SubPictures photos={['http://lorempixel.com/1440/960/city', 'http://lorempixel.com/1440/960/city']} tinyPhotos={['http://lorempixel.com/60/40/city', 'http://lorempixel.com/60/40/city']} />);

    const spy = jest.spyOn(component.instance(), 'updateDimensions');

    global.addEventListener('resize', spy);
    global.dispatchEvent(new Event('resize'));

    expect(spy).toHaveBeenCalled();
  });
});
