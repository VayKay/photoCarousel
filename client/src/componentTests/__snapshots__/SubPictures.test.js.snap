// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubPictures should render correctly 1`] = `
<div
  className="subPictureColumn"
  style={
    Object {
      "height": NaN,
      "left": "50%",
      "minHeight": NaN,
    }
  }
>
  <div
    className="imgHoverZoom"
  >
    <div
      className="entryOne"
    >
      <img
        alt=""
        className="subPicture blur"
        data-src="http://lorempixel.com/1440/960/city"
        height={NaN}
        id="subPictures"
        name="two"
        onClick={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        src="http://lorempixel.com/60/40/city"
        style={
          Object {
            "background": "url(http://lorempixel.com/60/40/city)",
            "minHeight": NaN,
            "opacity": undefined,
            "top": 0,
          }
        }
        width={256}
      />
    </div>
  </div>
  <div
    className="imgHoverZoom"
  >
    <div
      className="entryTwo"
    >
      <img
        alt=""
        className="subPicture blur"
        data-src="http://lorempixel.com/1440/960/city"
        height={NaN}
        id="subPictures"
        name="three"
        onClick={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        src="http://lorempixel.com/60/40/city"
        style={
          Object {
            "background": "url(http://lorempixel.com/60/40/city)",
            "minHeight": NaN,
            "opacity": undefined,
            "top": 0,
          }
        }
        width={256}
      />
    </div>
  </div>
</div>
`;
